otp4gb/routing.py

```diff
@@ -0,0 +1,267 @@
+from typing import Any, Optional, Union
+from urllib import parse
+
+# Third party imports
+import pydantic
+import requests
+from shapely import geometry
+
+# Local imports
+
+##### CONSTANTS #####
+LOG = logging.getLogger(__name__)
+ROUTER_API_ROUTE = "otp/routers/default/plan"
+REQUEST_TIMEOUT = 5
+REQUEST_RETRIES = 10
+OTP_ERRORS = {"NO_TRANSIT": 406, "TOO_CLOSE": 409}
+
+##### CLASSES #####
+class Mode(enum.StrEnum):
+    TRANSIT = "TRANSIT"
+    BUS = "BUS"
+    RAIL = "RAIL"
+    TRAM = "TRAM"
+    WALK = "WALK"
+    BICYCLE = "BICYCLE"
+
+    @staticmethod
+    def transit_modes() -> set[Mode]:
+        return {Mode.TRANSIT, Mode.BUS, Mode.RAIL, Mode.TRAM}
+
+
+class RoutePlanParameters(pydantic.BaseModel):
+    fromPlace: geometry.Point
+    toPlace: geometry.Point
+    date: datetime.date
+    time: datetime.time
+    mode: Union[list[str], str]
+    arriveBy: bool = False
+    searchWindow: Optional[int] = None
+    wheelchair: bool = False
+    debugItineraryFilter: bool = False
+    showIntermediateStops: bool = False
+    maxWalkDistance: float = 1000
+    locale: str = "en"
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def params(self) -> dict[str, Any]:
+        params: dict[str, Any] = {
+            k: str(v) for k, v in self.dict().items() if v is not None
+        }
+
+        # Update any parameters which require specific creation
+        params.update(
+            {
+                "fromPlace": [str(i) for i in self.fromPlace.coords],
+                "toPlace": [str(i) for i in self.toPlace.coords],
+                "time": self.time.strftime("%I:%M%p"),
+                "date": self.date.strftime("%m-%d-%Y"),  # date in USA format
+                "mode": ",".join([s.upper() for s in self.mode]),
+            }
+        )
+        return params
+
+    @pydantic.validator("fromPlace", "toPlace", pre=True)
+    def _validate_point(cls, value) -> geometry.Point:
+        if isinstance(value, geometry.Point):
+            return value
+
+        if isinstance(value, str):
+            match = re.match(
+                r"^[(\s]*"  # Optional starting bracket
+                r"([-+]?\d+(?:\.\d+)?)[,\s]+"  # 1st int or float with separator
+                r"([-+]?\d+(?:\.\d+)?)"  # 2nd number
+                r"(?:[,\s]+([-+]?\d+(?:\.\d+)?))?"  # optional 3rd number with separator
+                r"[\s)]*$",  # optional ending bracket
+                value.strip(),
+            )
+            if match is None:
+                raise ValueError(f"invalid point value string '{value}'")
+
+            value = [n for n in match.groups() if n is not None]
+
+        if isinstance(value, (tuple, list)):
+            if len(value) in (2, 3):
+                return geometry.Point(*[float(x) for x in value])
+            raise ValueError(f"point should have 2 coordinates not {len(value)}")
+
+        raise TypeError(
+            f"expected collection or string of coordinates for point not '{value}'"
```

Comment:
need `type(value)` rather than `value` ?

Reply:
Fixed in 73014cf3fff0ce7ecfc1b0144123aafe469d019c